build-system/app-index/test/test-amphtml-helpers.js

```diff
@@ -0,0 +1,411 @@
+
+        const {firstElementChild} = root;
+        expect(firstElementChild.tagName).to.equal('SCRIPT');
+        expect(firstElementChild.getAttribute('type'))
+            .to.equal('application/json');
+      });
+
+      it('renders json object', () => {
+        const id = 'whatever';
+        const state = {foo: 'bar', baz: {yes: 'no'}};
+
+        const {textContent} =
+            parseHtmlChunk(AmpState(id, state)).firstElementChild;
+
+        expect(JSON.parse(textContent)).to.deep.equal(state);
+      });
+
+      it('renders string literal', () => {
+        const id = 'whatever';
+        const state = 'foo';
+
+        const {textContent} =
+            parseHtmlChunk(AmpState(id, state)).firstElementChild;
+
+        expect(JSON.parse(textContent)).to.equal(state);
+      });
+
+      it('renders array', () => {
+        const id = 'whatever';
+        const state = ['foo', 'bar', 'baz'];
+
+        const {textContent} =
+            parseHtmlChunk(AmpState(id, state)).firstElementChild;
+
+        expect(JSON.parse(textContent)).to.deep.equal(state);
+      });
+    });
+
+    describe('addRequiredExtensionsToHead', () => {
+
+      it('renders ok', () => {
+        const rawStr = html`
+          <html>
+            <head></head>
+            <body>
+              <amp-foo foo=bar></amp-foo>
+            </body>
+          </html>`;
+
+        expect(new JSDOM(addRequiredExtensionsToHead(rawStr))).to.be.ok;
+      });
+
+      it('adds mixed', () => {
+
+        const expectedExtensions = [
+          'amp-foo',
+          'amp-bar',
+          'amp-foo-bar-baz',
+          'amp-bind',
+          'amp-form',
+        ];
+
+        const expectedTemplates = ['amp-mustache'];
+
+        const rawStr = html`
+          <html>
+            <head></head>
+            <body>
+              <amp-foo foo=bar></amp-foo>
+              <amp-foo foo=bar></amp-foo>
+              <amp-foo foo=bar></amp-foo>
+              <div>
+                <amp-bar></amp-bar>
+                <div>
+                  <amp-foo-bar-baz many=1 attributes=2>
+                    Text content
+                  </amp-foo-bar-baz>
+                </div>
+                <input>
+                <amp-state id=myState></amp-state>
+                <template type="amp-mustache"></template>
+              </div>
+            </body>
+          </html>`;
+
+        const {document} =
+            (new JSDOM(addRequiredExtensionsToHead(rawStr))).window;
+
+        const scripts =
+            Array.from(document.head.getElementsByTagName('script'));
+
+        expect(scripts.length)
+            .to.equal(expectedExtensions.length + expectedTemplates.length);
+
+        scripts.forEach(script => {
+          expect(script.getAttribute('src')).to.be.ok;
+          expect(script.getAttribute('async')).to.equal('');
+        });
+
+        expectedExtensions.forEach(expectedScript => {
+          expect(scripts.some(s => {
```

Comment:
If this assertion fails, it is a bit hard to know what went wrong. How about something like:

```javascript
expect(scripts).to.satisfy((arr) => {
  return arr.some((s) => {
    s.getAttribute('custom-element') == expectedScript;
    s.getAttribute('custom-template') == null;
  });
);
```

or

```javascript
const scriptEls = scripts.filter(s => s.getAttribute('custom-element') == expectedScript);
expect(scriptEls).to.have.length(1); // or not.be.empty if you want to check for more than one?
expect(scriptEls[0]).getAttribute('custom-template')).to.be.null;